From 5a216bacf475322d613708929dca3b090dfb8185 Mon Sep 17 00:00:00 2001
From: Makarand R Muley <makarand.r.muley@intel.com>
Date: Mon, 16 Nov 2015 22:25:04 +0530
Subject: [PATCH 1/7] FOR_UPSTREAM [VPG]: drm/i915: Limit max resolution of
 CHV-T3 to 2k1k

CHV-T3 supports maximum resolution of 2k1k@30fps on HDMI 1.4
This patch prunes all higher resolutions on T3 and enforces modes
which are no larger than 2k1k. This will prevent HDMI panels with
preferred resolution higher than 2k1k from blanking out on T3.

Change-Id: Ia6e755a457c4bd1326ad97e6d71f2b48f6ba65bb
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-11860
Signed-off-by: Makarand R Muley <makarand.r.muley@intel.com>
Reviewed-on: https://android.intel.com:443/446359
---
 drivers/gpu/drm/i915/i915_drv.h   | 4 ++++
 drivers/gpu/drm/i915/intel_hdmi.c | 7 ++++++-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 5c6085e..9adda04 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2553,6 +2553,10 @@ enum context_submission_status {
 
 #define GT_FREQUENCY_MULTIPLIER 50
 
+/* platform details */
+#define PCI_CHV_REV_ID_PACKAGE_TYPE_MASK	0x3
+#define PCI_CHV_REV_ID_PACKAGE_TYPE_T3		0x2
+
 #include "i915_trace.h"
 
 extern const struct drm_ioctl_desc i915_ioctls[];
diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index c0b125b..c3c9736 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -882,10 +882,15 @@ static void intel_disable_hdmi(struct intel_encoder *encoder)
 static int hdmi_portclock_limit(struct intel_hdmi *hdmi, bool respect_dvi_limit)
 {
 	struct drm_device *dev = intel_hdmi_to_dev(hdmi);
+	struct drm_i915_private *dev_priv = dev->dev_private;
+	bool is_chv_t3 = IS_CHERRYVIEW(dev) &&
+			((dev_priv->dev->pdev->revision &
+				PCI_CHV_REV_ID_PACKAGE_TYPE_MASK) ==
+					PCI_CHV_REV_ID_PACKAGE_TYPE_T3);
 
 	if ((respect_dvi_limit && !hdmi->has_hdmi_sink) || IS_G4X(dev))
 		return 165000;
-	else if (IS_HASWELL(dev) || INTEL_INFO(dev)->gen >= 8)
+	else if (IS_HASWELL(dev) || (INTEL_INFO(dev)->gen >= 8 && !is_chv_t3))
 		return 300000;
 	else
 		return 225000;
-- 
1.9.1

